kong/init.lua

```diff
@@ -1155,4 +1159,39 @@ function Kong.serve_cluster_listener(options)
 end
 
 
+do
+  local declarative = require("kong.db.declarative")
+  local cjson = require("cjson.safe")
+
+  function Kong.stream_config_listener()
+    local sock = ngx.req.socket()
+    if not sock then
+      ngx_log(ngx_CRIT, "unable to obtain request socket")
+      return
+    end
+
+    local data, err = sock:receive("*a")
+    if err then
+      ngx_log(ngx_CRIT, "unable to receive new config: ", err)
+      return
+    end
+
+    local parsed = cjson.decode(data)
+
+    local ok, err = concurrency.with_worker_mutex({ name = "dbless-worker" }, function()
+      return declarative.load_into_cache_with_events(parsed[1], parsed[2])
+    end)
+
+    if err == "no memory" then
+      kong.log.err("not enough cache space for declarative config")
+      return
+    end
+
+    if not ok then
+      kong.log.err("failed loading declarative config into cache: ", err)
+    end
```

Comment:
Easier to read and less prone to forgetting a `return` here in the future should we add code below:
```lua
if not ok then
  if err == "no memory"
    kong.log.err("not enough cache space for declarative config")
  
  else
    kong.log.err("failed loading declarative config into cache: ", err)
  end

  return
end
```

Also, we should make the "no memory" error actionable - instruct users how to increase the cache space, i.e. which config property to tweak.